tests/v2/actions/projects/projects.go

```diff
@@ -0,0 +1,71 @@
+package projects
+
+import (
+	"strings"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/shepherd/clients/rancher"
+	management "github.com/rancher/shepherd/clients/rancher/generated/management/v3"
+	"github.com/rancher/shepherd/extensions/defaults"
+	"github.com/rancher/shepherd/extensions/kubeapi/namespaces"
+	projectsapi "github.com/rancher/shepherd/extensions/kubeapi/projects"
+	"github.com/rancher/shepherd/extensions/projects"
+	namegen "github.com/rancher/shepherd/pkg/namegenerator"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kwait "k8s.io/apimachinery/pkg/util/wait"
+)
+
+// CreateProjectAndNamespace is a helper to create a project (norman) and a namespace in the project
+func CreateProjectAndNamespace(client *rancher.Client, clusterID string) (*management.Project, *corev1.Namespace, error) {
+	createdProject, err := client.Management.Project.Create(projects.NewProjectConfig(clusterID))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	namespaceName := namegen.AppendRandomString("testns")
+	projectName := strings.Split(createdProject.ID, ":")[1]
+
+	createdNamespace, err := namespaces.CreateNamespace(client, clusterID, projectName, namespaceName, "{}", map[string]string{}, map[string]string{})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return createdProject, createdNamespace, nil
+}
+
+// WaitForProjectFinalizerToUpdate is a helper to wait for project finalizer to update to match the expected finalizer count
+func WaitForProjectFinalizerToUpdate(client *rancher.Client, projectName string, projectNamespace string, finalizerCount int) error {
+	err := kwait.Poll(defaults.FiveHundredMillisecondTimeout, defaults.TenSecondTimeout, func() (done bool, pollErr error) {
+		project, pollErr := projectsapi.ListProjects(client, projectNamespace, metav1.ListOptions{
+			FieldSelector: "metadata.name=" + projectName,
+		})
+		if pollErr != nil {
+			return false, pollErr
+		}
+
+		if len(project.Items[0].Finalizers) == finalizerCount {
+			return true, nil
+		}
+		return false, pollErr
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UpdateProjectNamespaceFinalizer is a helper to update the finalizer in a project
+func UpdateProjectNamespaceFinalizer(client *rancher.Client, existingProject *v3.Project, finalizer []string) (*v3.Project, error) {
+	updatedProject := existingProject.DeepCopy()
+	updatedProject.ObjectMeta.Finalizers = finalizer
```

Comment:
yes, it depends on the calling function. F.E here, in this case, we are appending a new finalizer to the existing finalizers and then calling this function https://github.com/rancher/rancher/pull/45601/commits/bda05dc5b9edde2b128f14491c78bf0cad73a539#diff-e13c0ed52a4095e1e5f66be11a9f7ba2d0c2def7a5fa7dbbb47a328d8c1f6667L68

Reply:
oh so you are appending other finalizers before it goes to this function right? can we move the append part here? I ask because if this is used in other functions, without realizing someone can replace their finalizers with just the one that was sent in function statement.

Reply:
IMO we can keep this function as is. This allows us to add, remove, or overwrite the finalizer if needed. If we append it here, it won't be usable for overwrite scenarios. This approach is similar to our existing update functions, where we update an existing object (such as a project or global role) by passing the updates.

Reply:
can we not use the updateProject directly instead? by updating the dummy finalizer and sending it as a project? why use a separate function? thinking if we have this, should we be doing for each object similarly like annotations or labels?

Reply:
IIRC It was to avoid repeating these three lines each time I needed to update the finalizer:
```
updatedProject := existingProject.DeepCopy()
    updatedProject.ObjectMeta.Finalizers = finalizer

    updatedProject, err := projectsapi.UpdateProject(client, existingProject, updatedProject)
```
for this specific test suite: https://github.com/rancher/rancher/blob/release/v2.9/tests/v2/validation/projects/rbac_terminating_project_test.go

Reply:
Based on our discussion offline, keeping this as is